sw/device/tests/dif/dif_rstmgr_unittest.cc

```diff
@@ -0,0 +1,306 @@
+  EXPECT_EQ(dif_rstmgr_init(params, nullptr), kDifRstmgrBadArg);
+}
+
+TEST_F(InitTest, Success) {
+  dif_rstmgr_t rstmgr;
+  dif_rstmgr_params_t params = {.base_addr = dev().region()};
+  EXPECT_EQ(dif_rstmgr_init(params, &rstmgr), kDifRstmgrOk);
+}
+
+class ResetTest : public RstmgrTest {};
+
+TEST_F(ResetTest, NullArgs) {
+  EXPECT_EQ(dif_rstmgr_reset(nullptr), kDifRstmgrBadArg);
+}
+
+TEST_F(ResetTest, Success) {
+  EXPECT_WRITE32(RSTMGR_RESET_INFO_REG_OFFSET,
+                 std::numeric_limits<uint32_t>::max());
+  EXPECT_WRITE32(RSTMGR_SW_RST_CTRL_N_REG_OFFSET,
+                 std::numeric_limits<uint32_t>::max());
+
+  EXPECT_EQ(dif_rstmgr_reset(&rstmgr_), kDifRstmgrOk);
+}
+
+class ResetLockTest : public RstmgrTest {};
+
+TEST_F(ResetLockTest, NullArgs) {
+  EXPECT_EQ(dif_rstmgr_reset_lock(nullptr, 0), kDifRstmgrBadArg);
+}
+
+TEST_F(ResetLockTest, BadPeripheral) {
+  EXPECT_EQ(dif_rstmgr_reset_lock(&rstmgr_, RSTMGR_PARAM_NUMSWRESETS),
+            kDifRstmgrBadArg);
+}
+
+TEST_F(ResetLockTest, Success) {
+  for (uint32_t bit_index = 0; bit_index < RSTMGR_PARAM_NUMSWRESETS;
+       ++bit_index) {
+    uint32_t bitfield = bitfield_bit32_write(
+        std::numeric_limits<uint32_t>::max(), bit_index, false);
+    EXPECT_WRITE32(RSTMGR_SW_RST_REGEN_REG_OFFSET, bitfield);
+
+    EXPECT_EQ(dif_rstmgr_reset_lock(&rstmgr_, bit_index), kDifRstmgrOk);
+  }
+}
+
+class ResetIsLockedTest : public RstmgrTest {};
+
+TEST_F(ResetIsLockedTest, NullArgs) {
+  bool is_locked;
+  EXPECT_EQ(dif_rstmgr_reset_is_locked(nullptr, 0, nullptr), kDifRstmgrBadArg);
+  EXPECT_EQ(dif_rstmgr_reset_is_locked(nullptr, 0, &is_locked),
+            kDifRstmgrBadArg);
+  EXPECT_EQ(dif_rstmgr_reset_is_locked(&rstmgr_, 0, nullptr), kDifRstmgrBadArg);
+}
+
+TEST_F(ResetIsLockedTest, BadPeripheral) {
+  bool is_locked;
+  EXPECT_EQ(dif_rstmgr_reset_is_locked(&rstmgr_, RSTMGR_PARAM_NUMSWRESETS,
+                                       &is_locked),
+            kDifRstmgrBadArg);
+}
+
+TEST_F(ResetIsLockedTest, Success) {
+  for (uint32_t bit_index = 0; bit_index < RSTMGR_PARAM_NUMSWRESETS;
+       ++bit_index) {
+    uint32_t bit_locked = bitfield_bit32_write(
+        std::numeric_limits<uint32_t>::max(), bit_index, false);
+    EXPECT_READ32(RSTMGR_SW_RST_REGEN_REG_OFFSET, bit_locked);
+
+    bool is_locked = false;
+    EXPECT_EQ(dif_rstmgr_reset_is_locked(&rstmgr_, bit_index, &is_locked),
+              kDifRstmgrOk);
+    EXPECT_TRUE(is_locked);
+
+    is_locked = true;
+    EXPECT_READ32(RSTMGR_SW_RST_REGEN_REG_OFFSET, {{bit_index, true}});
+    EXPECT_EQ(dif_rstmgr_reset_is_locked(&rstmgr_, bit_index, &is_locked),
+              kDifRstmgrOk);
+    EXPECT_FALSE(is_locked);
+  }
+}
+
+class ResetCausesGetTest : public RstmgrTest {
+ protected:
+  ResetCausesGetTest() {
+    // Make sure that the test is up-to-date with the implementation.
+    EXPECT_EQ(0x1 << reset_info_reasons_.back(), kDifRstmgrResetInfoLast);
```

Comment:
Why not make the std::vector a constexpr std::array? Then this can be a static_assert (since std::array::back is constexpr).

Reply:
Will try that.

Reply:
I don't think that is a big issue. Changing that to `constexpr std::array` throws compilation errors elsewhere which I don't want to spend my time on.

Will keep it as it is.

Reply:
You need to define it out-of-band because C++14. This is fixed in C++17. Let's leave it like this for now.